script.js

```diff
@@ -966,64 +946,103 @@ require([
+    const disctinctSectionsArr = [...new Set(sectionIntArr)];
+    const sortedSections = disctinctSectionsArr.sort((a, b) => a-b).map(element => element.toString().padStart(2, '0'));
+
+    sortedSections.forEach(function (value) {
+      const option = domConstruct.create("option");
+      option.text = value
       sectionSelect.add(option);
     });
   }
 
-  // when township changes, reset the other dropdowns.
-  on(townshipSelect, "change", function (evt) {
-    var type = evt.target.value;
-    var i;
-    for (i = rangeSelect.options.length - 1; i >= 0; i--) {
-      rangeSelect.remove(i);
+  const handleResults = results => {
+    if (results.features.length > 0) {      
+      return results;
+    } else { 
+        $("#trs").prepend(
+          `<div id="TRSAlert" class="alert alert-danger" role="alert">
+            <button type="button" class="close" data-dismiss="alert" aria-label="Close"><span aria-hidden="true">&times;</span></button>
+            Invalid Township-Range combination. Please enter a correct Township-Range combination.
+          </div>`
+        )
+        window.setTimeout(function() {
+          $("#TRSAlert").fadeTo(500, 0).slideUp(500, function(){
+              $(this).remove(); 
+          });
+        }, 4000);
+        // remove sections
+        for (j = sectionSelect.options.length - 1; j >= 0; j--) {
+          sectionSelect.remove(j);
+        }
+        throw new Error('This is an invalid Township-Range combination');;
     }
+  }
 
-    var rangeQuery = new Query({
-      where: "twn_ch = '" + type.substr(0, 2) + "' AND tdir = '" + type.substr(2) + "'",
-      outFields: ["rng_ch", "rdir"],
-      returnDistinctValues: true,
-      orderByFields: ["rng_ch", "rdir"]
-    });
-    return townshipRangeSectionLayer.queryFeatures(rangeQuery).then(buildRangeDropdown);
-  })
-
-  // when range changes, reset the section dropdown.
-  on(rangeSelect, "change", function (evt) {
-    var type = evt.target.value;
-    var j;
-    for (j = sectionSelect.options.length - 1; j >= 0; j--) {
-      sectionSelect.remove(j);
+  function queryTR (type, whichDropdown) {
+    if(whichDropdown === 'range') { 
+      if (townshipSelect.value !== "Zoom to a Township") { // check to see if combo is valid if populated
+        const townshipValue = townshipSelect.value;
+        const TRQuery = new Query({
+          where: "rng_ch = '" + type.substr(0, 2) + "' AND rdir = '" + type.substr(2) + "' AND twn_ch = '" + townshipValue.substr(0, 2) + "' AND tdir = '" + townshipValue.substr(2) + "'",
+          returnGeometry: true,
+          outFields: ["*"]
+        });
+        townshipRangeSectionLayer.queryFeatures(TRQuery)
+        .then(results => handleResults(results))
+        .then(results => zoomToTRFeature(results))
+        .then(results => buildSectionDropdown(results))
+        .catch((error) => {
+          console.error(error);
+        });
```

Comment:
I like this idea. I'll do this. Can you go into a little bit more detail about your point about the arbitrary order? What do you suggest I do in order to accomplish this? An obvious choice would be to return results in buildSectionDropdown(). I chose to chain everything because it gave me a clear flow of work and because I could use handleResults() as a gatekeeper to go or not to go on with the rest of the chain.

Reply:
Okay, I see what you mean! :-D You're right, they're not dependent on eachother, so they shouldn't be linked together.